kratos/modeler/cad_tessellation_modeler.cpp

```diff
@@ -0,0 +1,356 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Dagmawi Bekel
+//                   Ruben Zorrilla
+//
+
+// Project includes
+#include "includes/define.h"
+#include "utilities/delaunator_utilities.h"
+#include "utilities/parallel_utilities.h"
+#include "cad_tessellation_modeler.h"
+
+namespace Kratos
+{
+///@name Operations
+///@{
+
+Modeler::Pointer CadTessellationModeler::Create(Model& rModel,
+    const Parameters ModelParameters) const
+{
+    return Kratos::make_shared<CadTessellationModeler>(rModel, ModelParameters);
+}
+
+///@}
+///@name Stages
+///@{
+
+void CadTessellationModeler::SetupModelPart()
+{
+    KRATOS_ERROR_IF_NOT(mParameters.Has("cad_model_part_name"))
+        << "Missing \"cad_model_part_name\" in CadTessellationModeler Parameters" << std::endl;
+    ModelPart& cad_model_part =
+        mpModel->GetModelPart(mParameters["cad_model_part_name"].GetString());
+
+    KRATOS_ERROR_IF_NOT(mParameters.Has("skin_model_part_name"))
+        << "Missing \"skin_model_part_name\" in CadTessellationModeler Parameters" << std::endl;
+    const std::string skin_model_part_name = mParameters["skin_model_part_name"].GetString();
+    ModelPart &skin_model_part = mpModel->HasModelPart(skin_model_part_name)
+                                     ? mpModel->GetModelPart(skin_model_part_name)
+                                     : mpModel->CreateModelPart(skin_model_part_name);
+
+    KRATOS_ERROR_IF(skin_model_part.NumberOfNodes() != 0)
+        << "Skin model part should be empty. Current number of nodes: "
+        << skin_model_part.NumberOfNodes() << std::endl;
+
+    IndexType node_id = 0;
+    IndexType element_id = 0;
+    IndexType vertex_id = 0;
+    const auto& r_geometries = cad_model_part.Geometries();
+    for (auto it = r_geometries.begin(); it != r_geometries.end(); ++it) {
+
+        IndexType trim_index = 0;
+        if (it->GetGeometryType() == GeometryData::Kratos_Brep_Surface) {
+
+            const auto& r_aux_geometry = *it;
+            const auto r_brep_surface_geom = dynamic_cast<const BrepSurfaceType&>(r_aux_geometry);
+
+            std::vector<array_1d<double, 2>> boundary_loop_uv;
+            IndexType point_id = 0;
+
+            while (r_brep_surface_geom.HasGeometryPart(trim_index)) {
+
+                auto p_surface_geometry = cad_model_part.pGetGeometry(r_brep_surface_geom.Id());
+                auto p_surface_trim = p_surface_geometry->pGetGeometryPart(trim_index);
+                auto p_brep_curve_on_surface =
+                    dynamic_pointer_cast<BrepCurveOnSurfaceType>(p_surface_trim);
+
+                auto boundary_segment_tessellation = this->ComputeBoundaryTessellation(*p_brep_curve_on_surface);
```

Comment:
there is another PR which avoids the cast and makes the tessellation on the 2D

Reply:
Can you point to that. I don't mind keeping the other one if it is more efficient.